### ISSUE TITLE
Method calls or subs intermittently don't receive their data

### ISSUE THREAD

**Issue by derwaldgeist** [meteor/meteor#11298]
I just posted this issue on the Meteor forums:
https://forums.meteor.com/t/strange-phenomenon-subsequent-methods-or-subs-wont-succeed/55029
and peterfkrueger was to kind to point me to https://github.com/meteor/meteor/issues/7868

I am facing a similar problem with Meteor 1.10.2, but mine is not only affecting the callbacks of methods, but also subs, so I thought it would make sense to open a new issue instead of reviving this super-old one.

To re-phrase what I wrote in the forums: Every once in a while, my users can execute one task in our app, which means logging them in and subscribing to an initial sub, but after that, all subsequent methods or subs fail. Methods don't get their callbacks fired, and subs don't receive their data (i.e. `withTracker()` won't re-run). In effect, the UI gets stuck and loading spinners spin forever, until I reload the browser.

This affects all the platforms we support, i.e. Desktop (tested on Chrome under macOS), Android, and iOS. For the latter, we are using a web-view embedded inside a Unity application. The problem here is that the users don't have a chance to reload the browser at all, so the app is drop-dead from that moment on.

The problem completely goes away once I restart the server (which is running on AWS EC2, using Mongo DB Atlas as its backend). Once the server has been restarted, everything works as it should. Until the problem re-appears after some days or weeks.

I noted three occurrences of this during the past months. When it happened yesterday, I used the Desktop browser and the Safari inspector to track down the problem. However, Meteor still reported the connection as being connected. I also checked the Monti APM logs for any irregularities, but couldn't find any. The problem doesn't show up on my testing server, nor is it reproducible in the dev environment. It seems as if it takes some days or weeks until it shows up. The only errors I saw in the server logs were failed Facebook auth logins. On client-side, there are no logs or timeout messages whatsoever. The calls and subs just silently fail.

In the issue linked above, people are suggesting to update `npm-mongo`, but this seems to have been fixed in Meteor itsef a while ago (the issue dates back to 2016), so I would rule that out. Another suggestion is to check unresolved promises. I will do that, but I am pretty sure that my server isn't using any, at least if a desktop browser is being used.

**Comment by jamesgibson14** [comment]
I have had the same issue with subs. There were no errors anywhere but the client never got any data. The publication queried the data correctly and added it to the sub, but the data never went through to the client.

**Comment by filipenevola** [comment]
Hi @derwaldgeist, are you handling errors correctly?

Also, is your code unblocking everything that is possible?

I already saw these problems happening when you have something that is never finishing, like a promise, then if you are not using unblocked publications and methods your next calls are not going to be executed.

Very often you don't need to block your methods and publications, to be honest I never block my methods and publications.

I have this mixin on all my methods:
```js
export default options => {
  const { run } = options;

  return assign(options, {
    run(...args) {
      this.unblock();
      return run.call(this, ...args);
    },
  });
};
```

I also unblock all my publications with this package https://github.com/lamhieu-vk/meteor-unblock and this fork https://github.com/pathable/meteor-publish-composite with this setting in case of publishComposite:

```json
  "publishComposite": {
    "unblock": true
  },
```

Unblocking everything is not the fix for the root cause, that is probably something in one of your publications or methods, but at least is going to prevent any side effects.

I don't believe there are many cases were block is necessary, I never found one in more than 7 years 😉 

You can read more about `unblock` [here](https://galaxy-guide.meteor.com/apm-managing-waittime.html#Using-this-unblock)

Do you see any indication of a bug in a Meteor core package?

**Comment by larswolter** [comment]
Hi,

We also had this problem in one of our applications, and we found out it was a combination of timing and a simple error. We had a publication that looks like the following:
```javascript
Meteor.publish('data',function(){
  if(this.userId) return collection.find();
})
```

When this publication is called without being logged in (which only happened in edge cases), it never resolves, because no cursor is returned or this.ready() is not called. All other calls and subscriptions are now waiting.
So for us it was to check if all subscriptions return a cursor or empty Array or do this.ready() in all possible cases.

Greetings
Lars

**Comment by derwaldgeist** [comment]
@filipenevola As already written in the forum thread, feel free to close the issue. I added unblocks to all my methods now. I am still a bit skeptical if this will solve the issue, mainly because it appeared immediately on the initial browser load, and I am pretty sure there is no method involved here that might block. But I'll keep an eye on my prod server and report back in case the problem re-appears.

@larswolter Unsure if this is really related, because this would block all the times. In my case, the problem only appeared after some weeks of normal operation, and once I restarted the server, it disappeared (at least for another couple of weeks).

**Comment by filipenevola** [comment]
Ok, closing the issue.

Feel free to comment here if necessary 😉

**Comment by santiagopuentep** [comment]
@filipenevola maybe the publication's `this.unblock()` should be a core feature instead of a package?

The APM documentation recommends https://github.com/meteorhacks/unblock, but you are using a different one. Is one preferable to the other?

**Comment by santiagopuentep** [comment]
I just saw that `meteorhacks:unblock` is no longer maintained.

**Comment by derwaldgeist** [comment]
@filipenevola recommended this fork: https://github.com/lamhieu-vk/meteor-unblock, but it is unmaintained as well.

**Comment by derwaldgeist** [comment]
Just from my personal experience: After adding unblocking to my subs, their response times got significantly *slower*, which is contrary to what I would have expected. I don't know why this happens, but it was reproducible.

**Comment by filipenevola** [comment]
Hi @derwaldgeist,

> @filipenevola recommended this fork: lamhieu-vk/meteor-unblock, but it is unmaintained as well.

Did you get any issues with https://github.com/lamhieu-vk/meteor-unblock? We can republish it under a [verified account](https://atmospherejs.com/i/faq) if you need to fix things, at least in my use cases it's working as expected so I don't see a reason to fork it now.

> which is contrary to what I would have expected. I don't know why this happens, but it was reproducible.

Maybe this is because now your server is not running each task sequentially and then the event loop is getting other tasks during the execution of a larger task?

If you were expecting a difference in performance for better just because you are unblocking the execution of parallel tasks I don't think this is what you are going to get. Each task will take exact the same time or more, if you start to run multiple things in "parallel".

**Comment by derwaldgeist** [comment]
> Did you get any issues with https://github.com/lamhieu-vk/meteor-unblock?

Nope, I just mentioned it because @santiagopuentep explicitly mentioned that the old meteorhacks version is unmaintained.

> Maybe this is because now your server is not running each task sequentially and then the event loop is getting other tasks during the execution of a larger task?

The tasks are sequential, but maybe the amount of documents sent to the client is causing this (if many requests come in in quick succession).

**Comment by derwaldgeist** [comment]
For anyone facing similar problems: I just found out that adding `this.unblock()` to all pubs is *not* the wisest idea, as it has unexpected side-effects:

https://forums.meteor.com/t/strange-phenomenon-subsequent-methods-or-subs-wont-succeed/55029/68?u=waldgeist
